pkg/credentialprovider/plugin/framework/plugin.go

```diff
@@ -0,0 +1,152 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/credentials/install"
+	"k8s.io/kubernetes/pkg/credentialprovider/apis/credentials/v1alpha1"
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+func init() {
+	install.Install(scheme)
+}
+
+// CredentialProvider is an interface implemented by the kubelet credential provider plugin to fetch
+// the username/password based on the provided image name.
+type CredentialProvider interface {
+	GetCredentials(ctx context.Context, image string, args []string) (response *v1alpha1.CredentialProviderResponse, err error)
+}
+
+// ExecPlugin implements the exec-based plugin for fetching credentials that is invoked by the kubelet.
+type ExecPlugin struct {
+	plugin CredentialProvider
+}
+
+// NewCredentialProvider returns an instance of execPlugin that fetches
+// credentials based on the provided plugin implementing the CredentialProvider interface.
+func NewCredentialProvider(plugin CredentialProvider) *ExecPlugin {
+	return &ExecPlugin{plugin}
+}
+
+// Run executes the credential provider plugin. Required information for the plugin request (in
+// the form of v1alpha1.CredentialProviderRequest) is provided via stdin from the kubelet.
+// The CredentialProviderResponse, containing the username/password required for pulling
+// the provided image, will be sent back to the kubelet via stdout.
+func (e *ExecPlugin) Run(ctx context.Context) error {
+	return e.runPlugin(ctx, os.Stdin, os.Stdout, os.Args[1:])
+}
+
+func (e *ExecPlugin) runPlugin(ctx context.Context, r io.Reader, w io.Writer, args []string) error {
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return err
+	}
+
+	gvk, err := json.DefaultMetaFactory.Interpret(data)
+	if err != nil {
+		return err
+	}
+
+	if gvk.GroupVersion() != v1alpha1.SchemeGroupVersion {
+		return fmt.Errorf("group version %s is not supported", gvk.GroupVersion())
+	}
+
+	request, err := decodeRequest(data)
+	if err != nil {
+		return err
+	}
+
+	if request.Image == "" {
+		return errors.New("image in plugin request was empty")
+	}
+
+	response, err := e.plugin.GetCredentials(ctx, request.Image, args)
+	if err != nil {
+		return err
+	}
+
+	if response == nil {
+		return errors.New("CredentialProviderResponse from plugin was nil")
+	}
+
+	encodedResponse, err := encodeResponse(response)
+	if err != nil {
+		return err
+	}
+
+	writer := bufio.NewWriter(w)
+	defer writer.Flush()
```

Comment:
if we're only doing a single Write call, I'm not sure a buffer here gets you much

Reply:
Resolving since I removed the framework package which mostly served as a reference implementation.